lib/sles4sap_publiccloud_basetest.pm

```diff
@@ -0,0 +1,56 @@
+# SUSE's openQA tests
+#
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+# Maintainer: QE-SAP <qe-sap@suse.de>
+# Summary: Base test providing cleanup, post fail and post run hooks for qe-sap-deployment project.
+# https://github.com/SUSE/qe-sap-deployment
+
+package sles4sap_publiccloud_basetest;
+
+use Mojo::Base 'publiccloud::basetest';
+use strict;
+use warnings FATAL => 'all';
+use Exporter 'import';
+use testapi;
+use qesapdeployment;
+
+our @EXPORT = qw(cleanup);
+
+sub cleanup {
+    my ($self) = @_;
+    # Do not run destroy if already executed
+    return if ($self->{cleanup_called});
+    $self->{cleanup_called} = 1;
+    my $inventory_check_cmd = join(" ", ("test", "-f", qesap_get_inventory()));
+    if (script_run($inventory_check_cmd) == 0) {
+        record_info("Ansible cleanup");
+        my $ansible_cleanup_rc = qesap_execute(verbose => "--verbose", cmd => "ansible", cmd_options => "-d", timeout => 1200);
+        force_soft_failure("Ansible destroy failed.") if $ansible_cleanup_rc != 0;
```

Comment:
What is the point of this soft fail ?
I would rather do cleanup again.
Nobody is looking at soft fails if there is no reason.

Reply:
Yeah... that is true :/
The idea was that if ansible cleanup fails for some reason, the test would continue with terraform cleanup and destroy at least infrastructure/resources.
But what if cleanup fails for the second time again? Should the whole test fail as well? 

I see now I have the same thing with terraform destroy fail. That is not a good idea since nobody would cleanup the resources manually.

Reply:
I would make test fail, it will be better at begging.
We will see that there are some issues even after retry and debug it.
Otherwise there will be left resources. :money_mouth_face:

Reply:
There is one thing I am afraid about ... the test will fail multiple times on ansible cleanup, and won't proceed to terraform destroy.
A person reviewing the test will work on the issue but will not think about going and manually do the cleanup. 
Is it possible to have:
- ansible cleanup will fail
- test will be set as failed, but will continue to terraform destroy, so the resources are forced to be cleaned up?

Reply:
I don't know how often tend ansible destroy to fail.
We should focus to make sure the resource is destroyed.
Like retry (or third retry) could be done with terraform destroy ?

Reply:
Yep that makes sense. 
So something like 3x retry for both ansible and terraform and fail afterwards?

Reply:
Sure :+1:

Reply:
> There is one thing I am afraid about ... the test will fail multiple times on ansible cleanup, and won't proceed to terraform destroy.

Shouldn't `lib/publiccloud/provider` call the `cleanup()` method on failures which issues a `terraform destroy`?

Reply:
Yes, but we need the ansible destroy as well so it for example de-registers host at the end. 
This one also uses the qe-sap-deployment projects glue script instead of plain commands.